### PR TITLE
ECL-402 - Fixed validation of CT UTR form for UA Entity Type

### PR DESCRIPTION
**Pull request by nemanjatomanovic**

### CODE REVIEW

File: conf/messages.en [hunk 1]
```diff
@@ -190,7 +190,7 @@ otherEntityType.ctutr.title = What is your Corporation Tax Unique Taxpayer Refer
 otherEntityType.ctutr.h1 = What is your Corporation Tax Unique Taxpayer Reference?
 otherEntityType.ctutr.body = Your Corporation Tax (CT) Unique Taxpayer Reference (UTR) number is 10 digits long.
 otherEntityType.ctutr.hint = For example, 0123456789
-otherEntityType.ctutr.error.required = Enter a Corporation Tax (CT) Unique Taxpayer Reference (UTR)
+otherEntityType.addCtutr.error.required = Enter a Corporation Tax (CT) Unique Taxpayer Reference (UTR)
```
Comment (peter-hazell) [inline conf/messages.en:193]: why change this key?

File: app/uk/gov/hmrc/economiccrimelevyregistration/forms/mappings/Constraints.scala [hunk 2]
```diff
@@ -121,4 +121,11 @@ trait Constraints {
       Invalid(invalidKey, length)
     case _                               => Valid
   }
+
+  protected def areAllElementsNumbers(invalidKey: String): Constraint[String] = Constraint {
+    case input if !(input forall Character.isDigit) =>
+      Invalid(invalidKey)
+    case _                                          => Valid
+
+  }
```
Comment (peter-hazell) [inline app/uk/gov/hmrc/economiccrimelevyregistration/forms/mappings/Constraints.scala:130]:
```suggestion
  protected def onlyContainsNumbers(invalidKey: String): Constraint[String] = Constraint {
    case input if input.exists(!_.isDigit) =>
      Invalid(invalidKey)
    case _                                          => Valid
  }
```

This is also missing a unit test in `ConstraintsSpec`

File: conf/messages.en (hunk 1 above)
Comment (nemanjatomanovic) [reply conf/messages.en:193]: There was a duplicate. We still have `otherEntityType.ctutr.error.required` and is used when we are asking user does he have a CTUTR. "New" key is used in validation of form when user needs to type in a CTUTR.

Comment (peter-hazell) [reply conf/messages.en:193]: ok that makes sense, probably worth making the other message keys for this page consistent then e.g. `...addCtutr...` ?

Comment (nemanjatomanovic) [reply conf/messages.en:193]: Done :)

File: app/uk/gov/hmrc/economiccrimelevyregistration/forms/mappings/Constraints.scala (hunk 2 above)
Comment (nemanjatomanovic) [reply app/uk/gov/hmrc/economiccrimelevyregistration/forms/mappings/Constraints.scala]: Done & Done :)

**Review by peter-hazell — APPROVED** [review]

File: app/uk/gov/hmrc/economiccrimelevyregistration/forms/CtUtrFormProvider.scala [hunk 3]
```diff
@@ -24,7 +24,8 @@ class CtUtrFormProvider extends Mappings {
 
   def apply(): Form[String] =
     Form(
-      "value" -> text("otherEntityType.ctutr.error.required")
+      "value" -> text("otherEntityType.addCtutr.error.required")
         .verifying(exactLength(CTUTRLength, "otherEntityType.ctutr.error.length"))
+        .verifying(areAllElementsNumbers("otherEntityType.ctutr.error.length"))
```
[the remark this answers is not in the document]
Comment (nemanjatomanovic) [reply app/uk/gov/hmrc/economiccrimelevyregistration/forms/CtUtrFormProvider.scala]: Resolved on Slack :)

**Pull request merged**